core/src/main/scala/org/apache/spark/scheduler/Stage.scala

```diff
@@ -76,6 +76,37 @@ private[spark] abstract class Stage(
    */
   private var _latestInfo: StageInfo = StageInfo.fromStage(this, nextAttemptId)
 
+  /**
+   * Spark is resilient to executors dying by retrying stages on FetchFailures. Here, we keep track
+   * of unique stage failures (per stage attempt) triggered by fetch failures to prevent endless
+   * stage retries. Specifically, per stage we wish to only record a failure when the following
+   * holds:
+   *
+   * A) A fetch failure was observed
+   * B) A failure has not yet been registered for this stage attempt. There may be multiple
+   * concurrent failures for a sinlge stage since we may have multiple tasks executing at the same
+   * time, one or many of which may fail. Also, even though there may only be one non-zombie stage
+   * attemp, zombie stages may still have running tasks.
+   */
+  private val attemptsFailedFromFetch = new HashSet[Int]
+
+  private[scheduler] def clearFailures() : Unit = {
+    attemptsFailedFromFetch.clear()
+  }
+
+  /**
+   * Check whether we should abort the failedStage due to multiple failures.
+   * This method updates the running set of failures for a particular stage and returns
```

Comment:
can you change this to " ... running set of failed stage attempts ...", since its not clear if "failure" refers to task failures or stage attempt failures